component/loki/source/aws_firehose/component.go

```diff
@@ -0,0 +1,173 @@
+}
+
+// Component is the main type for the `loki.source.awsfirehose` component.
+type Component struct {
+	// mut controls concurrent access to fanout
+	mut    sync.RWMutex
+	fanout []loki.LogsReceiver
+
+	// destination is the main destination where the TargetServer writes received log entries to
+	destination loki.LogsReceiver
+	rbs         []*relabel.Config
+
+	server *fnet.TargetServer
+
+	opts component.Options
+	args Arguments
+
+	// utils
+	serverMetrics *util.UncheckedCollector
+	logger        log.Logger
+}
+
+// New creates a new Component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts:          o,
+		destination:   make(loki.LogsReceiver),
+		fanout:        args.ForwardTo,
+		serverMetrics: util.NewUncheckedCollector(nil),
+
+		logger: log.With(o.Logger, "component", "aws_firehose_logs"),
+	}
+
+	o.Registerer.MustRegister(c.serverMetrics)
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run starts a routine forwards received message to each destination component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		c.shutdownServer()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.destination:
+			c.mut.RLock()
+			for _, receiver := range c.fanout {
+				receiver <- entry
+			}
+			c.mut.RUnlock()
+		}
+	}
+}
+
+// Update updates the component with a new configuration, restarting the server if needed.
+func (c *Component) Update(args component.Arguments) error {
+	var err error
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	c.fanout = newArgs.ForwardTo
+
+	var newRelabels []*relabel.Config = nil
+	var relabelRulesChanged = false
+
+	// todo(pablo): is it a good practice to keep a reference to the arguments in the
+	// component struct, used for comparing here rather than destructuring them?
+	if newArgs.RelabelRules != nil && len(newArgs.RelabelRules) > 0 {
+		relabelRulesChanged = true
+		newRelabels = flow_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelRules)
+	} else if c.rbs != nil && len(c.rbs) > 0 && (newArgs.RelabelRules == nil || len(newArgs.RelabelRules) == 0) {
+		// nil out relabel rules if they need to be cleared
+		relabelRulesChanged = true
+	}
+
+	// Since the handler is created ad-hoc for the server, and the handler depends on the relabels
+	// consider this as a cause for server restart as well. Much simpler than adding a lock on the
+	// handler and doing the relabel rules change on the fly
+	serverNeedsUpdate := !reflect.DeepEqual(c.args.Server, newArgs.Server)
+	if !serverNeedsUpdate && !relabelRulesChanged {
+		c.args = newArgs
+		return nil
+	}
+
+	c.shutdownServer()
+
+	// update relabel rules in component if needed
+	if relabelRulesChanged {
+		c.rbs = newRelabels
+	}
+
+	jobName := strings.Replace(c.opts.ID, ".", "_", -1)
+
+	registry := prometheus.NewRegistry()
```

Comment:
`NewRegistry` feels off, is there a reason we arent using the passed in registry?

Reply:
Since both the component and the inner server register metrics, I'm using the unchecked collector registered when the component is created. If not, when either the server or the handler or the server re-created, the registry will try to re-register those metrics and panic. 
For the server there's no other way, all components that use that server use an unchecked collector.
For the handler metrics I could create the metrics object outside, and keep it in the component state. That way we avoid the re-registering panic

Reply:
The latter sounds reasonable and ensures the metrics get captured, if there isnt value to the metrics and your just using the registry to satisfy the function interface then thats a different store. If so then a comment explaining the behavior would suffice.

Reply:
Related: https://github.com/weaveworks/common/issues/285 - I will work on making sure we can get rid of the unchecked collector eventually.

Reply:
Implemented the latter to avoid passing the registry.